File: behavioral_patterns/template_method/typescript/src/BeverageProductionTemplate.ts
export abstract class BeverageProductionTemplate {
  private customerWantsCondiments: boolean = false;

  public prepareBeverage(): void {
      this.boilWater();
      this.mixIngredients();
      this.pourInCup();
      if (this.customerWantsCondiments) {
          this.addCondiments();
      }
  }

  protected boilWater(): void {
      console.log("Boiling water");
  }

  protected pourInCup(): void {
      console.log("Pouring into cup");
  }

  protected abstract mixIngredients(): void;
  protected abstract addCondiments(): void;

  public setCustomerWantsCondiments(wantsCondiments: boolean): void {
      this.customerWantsCondiments = wantsCondiments;
  }
}

File: behavioral_patterns/template_method/java/beverageProductionTemplate/src/template_method/BeverageProductionTemplate.java
package template_method;

public abstract class BeverageProductionTemplate {
    boolean customerWantsCondiments = false;

    public final void prepareBeverage() {
        boilWater();
        mixIngredients();
        pourInCup();
        if (customerWantsCondiments) {
            addCondiments();
        }
    }

    void boilWater() {
        System.out.println("Boiling water");
    }

    void pourInCup() {
        System.out.println("Pouring into cup");
    }

    abstract void mixIngredients();
    abstract void addCondiments();


    public void setCustomerWantsCondiments(boolean customerWantsCondiments) {
        this.customerWantsCondiments = customerWantsCondiments;
    }
}